firedrake/interpolation.py

```diff
@@ -53,26 +58,92 @@ class Interpolator(object):
        arguments (such that they won't be collected until the
        :class:`Interpolator` is also collected).
     """
-    def __init__(self, expr, V, subset=None, access=op2.WRITE):
-        self.callable = make_interpolator(expr, V, subset, access)
+    def __init__(self, expr, V, subset=None, freeze_expr=False, access=op2.WRITE):
+        self.callable, args = make_interpolator(expr, V, subset, access)
+        self.args = args
+        self.nargs = len(args)
+        self.freeze_expr = freeze_expr
+        self.V = V
 
     @utils.known_pyop2_safe
-    def interpolate(self):
+    def interpolate(self, *function, output=None, transpose=False):
         """Compute the interpolation.
 
+        :arg function: If the expression being interpolated contains an
+           :class:`ufl.Argument`, then the :class:`.Function` value to
+           interpolate.
+        :kwarg output: Optional. A :class:`.Function` to contain the output.
+        :kwarg transpose: Set to true to apply the transpose (adjoint) of the
+           interpolation operator.
+
         :returns: The resulting interpolated :class:`.Function`.
         """
-        return self.callable()
+        if transpose and not self.nargs:
+            raise ValueError("Cannot currently apply transpose interpolation with no argument.")
+        if self.nargs != len(function):
+            raise ValueError("Passed %d Functions to interpolate, expected %d"
+                             % (len(function), self.nargs))
+        try:
+            callable = self.frozen_callable
+        except AttributeError:
+            callable = self.callable()
+            if self.freeze_expr:
+                if self.nargs:
+                    self.frozen_callable = callable
+                else:
+                    self.frozen_callable = callable.copy()
+
+        if self.nargs:
+            callable._force_evaluation()
+            function, = function
+            if not transpose:
+                result = output or firedrake.Function(self.V)
+                with function.dat.vec_ro as x:
+                    with result.dat.vec_wo as out:
+                        callable.handle.mult(x, out)
+            else:
+                result = output or firedrake.Function(self.args[0].function_space())
+                with function.dat.vec_ro as x:
+                    with result.dat.vec_wo as out:
+                        callable.handle.multTranspose(x, out)
+
+            return result
+        else:
+            if output:
+                output.assign(callable)
+                return output
+            if isinstance(self.V, firedrake.Function):
+                self.V.assign(callable)
+                return self.V
```

Comment:
One of the points of the interpolator object (for reuse) is to avoid
unnecessary copies of temporary data and par-loops.
This looks to be a regression in that it requires that we write into a "temporary"
and then copy into self.V. Why could we not write into self.V straight away?

Reply:
This copy only occurs in the case of a frozen interpolation. In that case, it's the minimal work (the only thing that happens is we copy the frozen value).